pkg/distributor/distributor.go

```diff
@@ -1093,11 +1093,43 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
 		mimirpb.ReuseSlice(req.Timeseries)
 	})
 
-	err := d.PushWithMiddlewares(ctx, pushReq)
-	if err != nil {
-		return nil, err
+	pushErr := d.PushWithMiddlewares(ctx, pushReq)
+	if pushErr == nil {
+		return &mimirpb.WriteResponse{}, nil
 	}
-	return &mimirpb.WriteResponse{}, nil
+	handledErr := d.handlePushError(pushErr)
+	return nil, handledErr
+}
+
+func (d *Distributor) handlePushError(err error) error {
+	if errors.Is(err, context.Canceled) {
+		return err
+	}
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		switch distributorError := distributorError.(type) {
+		case distributor_error.ReplicasNotMatchDistributorPushError:
```

Comment:
Instead of checking specific types on errors, we should check `case errors.As(...):` as the error may come wrapped.

Reply:
In the current implementation it is not the case, but it might be that in the future this will happen.
I will add this check.